### PR TITLE
[ceres] upgrade to 2.0.0

### PR DESCRIPTION
**Pull request by fran6co**

### CODE REVIEW

**Comment by fran6co** [comment]
Theia needs to be fixed upstream to support ceres 2.0.0

File: ports/ceres/CONTROL [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Source: ceres
-Version: 1.14.0
-Port-Version: 10
+Version: 2.0.0
+Port-Version: 11
```
Comment (NancyLi1013) [inline ports/ceres/CONTROL:3]:
```suggestion
```
This field is no longer needed for a new update.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
The failures for `theia` on Linux:

```
/mnt/vcpkg-ci/installed/x64-linux/include/ceres/internal/integer_sequence_algorithm.h:64:21: error: ‘integer_sequence’ is not a member of ‘std’
 struct SumImpl<std::integer_sequence<T, N, Ns...>> {
                     ^~~~~~~~~~~~~~~~
/mnt/vcpkg-ci/installed/x64-linux/include/ceres/internal/integer_sequence_algorithm.h:64:21: error: ‘integer_sequence’ is not a member of ‘std’
/mnt/vcpkg-ci/installed/x64-linux/include/ceres/internal/integer_sequence_algorithm.h:64:46: error: wrong number of template arguments (3, should be 1)
 struct SumImpl<std::integer_sequence<T, N, Ns...>> {
                                              ^~~
```
Related issue from upstream https://github.com/ceres-solver/ceres-solver/issues/604

Needs to use C++ standard >= 14 to build.

**Comment by NancyLi1013** [comment]
Seems it doesn't work.

```
========== Warning: Compiling with Debug mode ==========
-- CMAKE_CXX_FLAGS: -fPIC -std=c++11
-- Found Eigen version 3.3.7: /mnt/vcpkg-ci/installed/x64-linux/include/eigen3
-- CMAKE_CXX_FLAGS: -fPIC -std=c++11
```

**Comment by fran6co** [comment]
> Seems it doesn't work.
> 
> ```
> ========== Warning: Compiling with Debug mode ==========
> -- CMAKE_CXX_FLAGS: -fPIC -std=c++11
> -- Found Eigen version 3.3.7: /mnt/vcpkg-ci/installed/x64-linux/include/eigen3
> -- CMAKE_CXX_FLAGS: -fPIC -std=c++11
> ```

@NancyLi1013 fixed now

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
@fran6co 

Have you test these features? Can they work fine?

**Comment by fran6co** [comment]
Yes, I have been using ceres 2 from master for awhile now. The api didn’t
get big changes

**Comment by NancyLi1013** [comment]
LGTM now, thanks for your PR @fran6co.

**Comment by strega-nil** [comment]
Cool, thanks @fran6co :)

**Comment by BillyONeal** [comment]
In CI we're seeing failures to build ceres probably resulting from this change:

```
cmd.exe /C "cd . && D:\downloads\tools\cmake-3.18.4-windows\cmake-3.18.4-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=internal\ceres\CMakeFiles\schur_eliminator_benchmark.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe" --manifests  -- "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x86\link.exe"  internal\ceres\CMakeFiles\schur_eliminator_benchmark.dir\schur_eliminator_benchmark.cc.obj  /out:bin\schur_eliminator_benchmark.exe /implib:lib\schur_eliminator_benchmark.lib /pdb:bin\schur_eliminator_benchmark.pdb /version:0.0 /machine:X86 /ignore:4049 /nologo    /debug /INCREMENTAL /subsystem:console  lib\ceres-debug.lib  D:\installed\x86-windows\debug\lib\benchmark.lib  D:\installed\x86-windows\debug\lib\glog.lib  D:\installed\x86-windows\debug\lib\gflags_debug.lib  shlwapi.lib  shlwapi.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK Pass 1: command "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x86\link.exe internal\ceres\CMakeFiles\schur_eliminator_benchmark.dir\schur_eliminator_benchmark.cc.obj /out:bin\schur_eliminator_benchmark.exe /implib:lib\schur_eliminator_benchmark.lib /pdb:bin\schur_eliminator_benchmark.pdb /version:0.0 /machine:X86 /ignore:4049 /nologo /debug /INCREMENTAL /subsystem:console lib\ceres-debug.lib D:\installed\x86-windows\debug\lib\benchmark.lib D:\installed\x86-windows\debug\lib\glog.lib D:\installed\x86-windows\debug\lib\gflags_debug.lib shlwapi.lib shlwapi.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:internal\ceres\CMakeFiles\schur_eliminator_benchmark.dir/intermediate.manifest internal\ceres\CMakeFiles\schur_eliminator_benchmark.dir/manifest.res" failed (exit code 1120) with the following output:
schur_eliminator_benchmark.cc.obj : error LNK2019: unresolved external symbol "public: __thiscall ceres::internal::BlockRandomAccessDenseMatrix::BlockRandomAccessDenseMatrix(class std::vector<int,class std::allocator<int> > const &)" (??0BlockRandomAccessDenseMatrix@internal@ceres@@QAE@ABV?$vector@HV?$allocator@H@std@@@std@@@Z) referenced in function "public: __thiscall ceres::internal::BenchmarkData::BenchmarkData(int)" (??0BenchmarkData@internal@ceres@@QAE@H@Z)
schur_eliminator_benchmark.cc.obj : error LNK2019: unresolved external symbol "public: __thiscall ceres::internal::BlockSparseMatrix::BlockSparseMatrix(struct ceres::internal::CompressedRowBlockStructure *)" (??0BlockSparseMatrix@internal@ceres@@QAE@PAUCompressedRowBlockStructure@12@@Z) referenced in function "public: __thiscall ceres::internal::BenchmarkData::BenchmarkData(int)" (??0BenchmarkData@internal@ceres@@QAE@H@Z)
schur_eliminator_benchmark.cc.obj : error LNK2019: unresolved external symbol "public: struct ceres::internal::CompressedRowBlockStructure const * __thiscall ceres::internal::BlockSparseMatrix::block_structure(void)const " (?block_structure@BlockSparseMatrix@internal@ceres@@QBEPBUCompressedRowBlockStructure@23@XZ) referenced in function "void __cdecl ceres::internal::BM_SchurEliminatorEliminate(class benchmark::State &)" (?BM_SchurEliminatorEliminate@internal@ceres@@YAXAAVState@benchmark@@@Z)
schur_eliminator_benchmark.cc.obj : error LNK2019: unresolved external symbol "public: __thiscall ceres::internal::ThreadPool::ThreadPool(void)" (??0ThreadPool@internal@ceres@@QAE@XZ) referenced in function "public: __thiscall ceres::internal::ContextImpl::ContextImpl(void)" (??0ContextImpl@internal@ceres@@QAE@XZ)
schur_eliminator_benchmark.cc.obj : error LNK2019: unresolved external symbol "public: __thiscall ceres::internal::ThreadPool::~ThreadPool(void)" (??1ThreadPool@internal@ceres@@QAE@XZ) referenced in function "public: virtual __thiscall ceres::internal::ContextImpl::~ContextImpl(void)" (??1ContextImpl@internal@ceres@@UAE@XZ)
schur_eliminator_benchmark.cc.obj : error LNK2019: unresolved external symbol "public: static class ceres::internal::SchurEliminatorBase * __cdecl ceres::internal::SchurEliminatorBase::Create(struct ceres::internal::LinearSolver::Options const &)" (?Create@SchurEliminatorBase@internal@ceres@@SAPAV123@ABUOptions@LinearSolver@23@@Z) referenced in function "void __cdecl ceres::internal::BM_SchurEliminatorEliminate(class benchmark::State &)" (?BM_SchurEliminatorEliminate@internal@ceres@@YAXAAVState@benchmark@@@Z)
bin\schur_eliminator_benchmark.exe : fatal error LNK1120: 6 unresolved externals
[156/162] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.18.4-windows\cmake-3.18.4-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=internal\ceres\CMakeFiles\small_blas_gemv_benchmark.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe" --manifests  -- "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x86\link.exe"  internal\ceres\CMakeFiles\small_blas_gemv_benchmark.dir\small_blas_gemv_benchmark.cc.obj  /out:bin\small_blas_gemv_benchmark.exe /implib:lib\small_blas_gemv_benchmark.lib /pdb:bin\small_blas_gemv_benchmark.pdb /version:0.0 /machine:X86 /ignore:4049 /nologo    /debug /INCREMENTAL /subsystem:console  lib\ceres-debug.lib  D:\installed\x86-windows\debug\lib\benchmark.lib  D:\installed\x86-windows\debug\lib\glog.lib  D:\installed\x86-windows\debug\lib\gflags_debug.lib  shlwapi.lib  shlwapi.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
[157/162] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.18.4-windows\cmake-3.18.4-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=internal\ceres\CMakeFiles\small_blas_gemm_benchmark.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe" --manifests  -- "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x86\link.exe"  internal\ceres\CMakeFiles\small_blas_gemm_benchmark.dir\small_blas_gemm_benchmark.cc.obj  /out:bin\small_blas_gemm_benchmark.exe /implib:lib\small_blas_gemm_benchmark.lib /pdb:bin\small_blas_gemm_benchmark.pdb /version:0.0 /machine:X86 /ignore:4049 /nologo    /debug /INCREMENTAL /subsystem:console  lib\ceres-debug.lib  D:\installed\x86-windows\debug\lib\benchmark.lib  D:\installed\x86-windows\debug\lib\glog.lib  D:\installed\x86-windows\debug\lib\gflags_debug.lib  shlwapi.lib  shlwapi.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
[158/162] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.18.4-windows\cmake-3.18.4-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=internal\ceres\autodiff_benchmarks\CMakeFiles\autodiff_benchmarks.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe" --manifests  -- "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x86\link.exe"  internal\ceres\autodiff_benchmarks\CMakeFiles\autodiff_benchmarks.dir\autodiff_benchmarks.cc.obj  /out:bin\autodiff_benchmarks.exe /implib:lib\autodiff_benchmarks.lib /pdb:bin\autodiff_benchmarks.pdb /version:0.0 /machine:X86 /ignore:4049 /nologo    /debug /INCREMENTAL /subsystem:console  lib\ceres-debug.lib  D:\installed\x86-windows\debug\lib\benchmark.lib  D:\installed\x86-windows\debug\lib\glog.lib  D:\installed\x86-windows\debug\lib\gflags_debug.lib  shlwapi.lib  shlwapi.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
   Creating library lib\autodiff_benchmarks.lib and object lib\autodiff_benchmarks.exp
   Creating library lib\autodiff_benchmarks.lib and object lib\autodiff_benchmarks.exp
[159/162] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.18.4-windows\cmake-3.18.4-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=internal\ceres\autodiff_benchmarks\CMakeFiles\autodiff_benchmarks_fast_math.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.19041.0\x64\mt.exe" --manifests  -- "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x86\link.exe"  internal\ceres\autodiff_benchmarks\CMakeFiles\autodiff_benchmarks_fast_math.dir\autodiff_benchmarks.cc.obj  /out:bin\autodiff_benchmarks_fast_math.exe /implib:lib\autodiff_benchmarks_fast_math.lib /pdb:bin\autodiff_benchmarks_fast_math.pdb /version:0.0 /machine:X86 /ignore:4049 /nologo    /debug /INCREMENTAL /subsystem:console  lib\ceres-debug.lib  D:\installed\x86-windows\debug\lib\benchmark.lib  D:\installed\x86-windows\debug\lib\glog.lib  D:\installed\x86-windows\debug\lib\gflags_debug.lib  shlwapi.lib  shlwapi.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
   Creating library lib\autodiff_benchmarks_fast_math.lib and object lib\autodiff_benchmarks_fast_math.exp
   Creating library lib\autodiff_benchmarks_fast_math.lib and object lib\autodiff_benchmarks_fast_math.exp
[160/162] "C:\PROGRA~2\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\bin\Hostx64\x86\cl.exe"   /TP -DCERES_BUILDING_SHARED_LIBRARY -DCERES_SUITESPARSE_VERSION=\"\" -DEIGEN_MPL2_ONLY -DGFLAGS_IS_A_DLL=1 -DGLOG_NO_ABBREVIATED_SEVERITIES -DGOOGLE_GLOG_DLL_DECL=__declspec(dllimport) -DGOOGLE_GLOG_DLL_DECL_FOR_UNITTESTS=__declspec(dllimport) -D_USE_MATH_DEFINES -D_VARIADIC_MAX=10 -ID:\buildtrees\ceres\src\2.0.0-e041eed1b2.clean\internal -ID:\buildtrees\ceres\src\2.0.0-e041eed1b2.clean\internal\ceres -Iconfig -ID:\buildtrees\ceres\src\2.0.0-e041eed1b2.clean\include -ID:\installed\x86-windows\include -ID:\installed\x86-windows\share\eigen3\..\..\include\eigen3 /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /GR /EHsc /MP  /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  /wd4018 /wd4267 /wd4099 /wd4996 /wd4800 /wd4244 /wd4251 /bigobj -std:c++14 /showIncludes /Fointernal\ceres\CMakeFiles\invert_psd_matrix_benchmark.dir\invert_psd_matrix_benchmark.cc.obj /Fdinternal\ceres\CMakeFiles\invert_psd_matrix_benchmark.dir\ /FS -c D:\buildtrees\ceres\src\2.0.0-e041eed1b2.clean\internal\ceres\invert_psd_matrix_benchmark.cc
ninja: build stopped: subcommand failed.
```

Any ideas what might be going on?

**Comment by BillyONeal** [comment]
I think it's:

```
if (BUILD_BENCHMARKS)
  find_package(benchmark QUIET)
  if (benchmark_FOUND)
     message("-- Found Google benchmark library. Building Ceres benchmarks.")
  else()
     message("-- Failed to find Google benchmark library, disabling build of benchmarks.")
     update_cache_variable(BUILD_BENCHMARKS OFF)
  endif()
  mark_as_advanced(benchmark_DIR)
endif()
```

and something being wrong with the benchmarks target. For now I'm going to pass `-DBUILD_BENCHMARKS=OFF`

**Pull request merged**